### ISSUE TITLE
Celery queues

### ISSUE THREAD

**Issue by rkorytkowski** [OpenConceptLab/ocl_issues#134]
Separate bulk import queue from other tasks. Create 2 new queues for imports with one being reserved for the root user.

Determine, if we need to guarantee the sequence of execution or we can execute tasks in parallel using multiple processes.

Make sure that a single user can queue multiple import tasks.

Return AlreadyQueued and do not queue an import, if the same import (same user, json and parameters) is requested, if the first one is still running.

Adjust the default queue to be able to run 20 tasks in parallel.

**Comment by paynejd** [comment]
At least for now I'm okay if we require that dependencies in bulk import requests must currently be managed by the client, not by OCL, so that we can run imports in parallel. I manage this in my own scripts now, which is why I've been focused on how the import results are provided.

How hard would it be for a user to submit an import request that:
1) Only ran after an import with task id xxxxxx was completed? (i.e. they provided a URL parameter with the dependency)
2) Or, add a URL parameter that indicated the import should be run in sequence for this user account?

**Comment by rkorytkowski** [comment]
1 and 2 are similarly complex as it's a significantly modified queue. There's no such logic out of the box in celery so I'd need to find a way to inject that logic. What is the actual use case? I'm trying to understand why the client cannot join the 2 imports and send them as one?

**Comment by paynejd** [comment]
The main reason I have needed to separate out imports is to get the mapping IDs. Mapping IDs are auto-generated but you need them if you are going to create an explicit reference to one of them. In scripts I've written, I typically follow this process:
1. Bulk import sources, concepts, mappings and source versions
2. Process the results of the import to get the mapping IDs
3. Bulk import collections, references (to concepts & mappings), and collection versions

Ticket #69 would implement the option to set a mapping ID manually, which would would allow me to import the above process as a single import request, assuming that I am able to manually assign IDs to mappings in the same way that I do for concepts.

Another situation would be when a user is editing metadata in a 3rd party app (like the PEPFAR country mapping app) and they hit submit, which sends the import request over to OCL. They then edit some more metadata and hit submit again. We would want to ensure that the new import is run after the initial import.

**Comment by rkorytkowski** [comment]
Deployed to QA. Waiting for https://issues.openmrs.org/browse/ITSM-4229 to be able to deploy it to other envs. 

I agree #69 is the right and easy solution for importing collections.

The 3rd party app could be solved with Celery chains I guess. Let's have a separate issue for that as it's a bigger one.

**Comment by paynejd** [comment]
@rkorytkowski Has this been deployed to all environments? Ready to close out?

**Comment by rkorytkowski** [comment]
Not yet. I'll use Cintia's workaround for ITSM-4229 and apply it to staging and production this week.

**Comment by rkorytkowski** [comment]
Deployed to all envs now, but routing tasks to proper queues does not seem to be working. I'll have another view.

**Comment by rkorytkowski** [comment]
Fixed.
